### ISSUE TITLE
Cannot read property 'build' of undefined

### ISSUE THREAD

**Issue by brianyoungblood** [ProboCI/probo#86]
I’ve made my way through the [quick start guide](https://github.com/ProboCI/probo/blob/master/QUICKSTART.md) and on the release/tag v3.3.2.  Github is touching my server by way of the webhook as well.

After I submit a pull request from my Github repo, stdout on the server running Probo returns an error "Cannot read property 'build' of undefined"

Here is a snippet of the error log:

```
{"name":"probo","hostname":"li171-190","pid":22686,"component":"http","req_id":"470bb3c0-8fe9-11e6-aad3-ffe2a13c958f","level":20,"event":"status updated","data":{"context":"ProboCI-local/env","status":{"state":"error","description":"[■] Cannot read property 'build' of undefined","context":"ProboCI-local/env","target_url":"http://55ed4e903fc00000.probo.nodesafe.com:3050/","user":"brianyoungblood","repo":"probo_wordpress","sha":"153584b716cd99a72f56fe9a9aa25e395c378da2"}},"msg":"build event for build 55ed4e903fc00000: status updated","time":"2016-10-11T19:31:16.614Z","src":{"file":"/opt/probo/lib/ContainerManager.js","line":40,"func":"emitBuildEvent"},"v":0}
{"name":"probo","hostname":"li171-190","pid":22686,"component":"http","req_id":"470bb3c0-8fe9-11e6-aad3-ffe2a13c958f","level":30,"status":{"state":"error","description":"[■] Cannot read property 'build' of undefined"},"msg":"status updated","time":"2016-10-11T19:31:16.615Z","src":{"file":"/opt/probo/lib/ContainerManager.js","line":320,"func":"updateStatus"},"v":0}
{"name":"probo","hostname":"li171-190","pid":22686,"component":"container-manager","level":50,"err":{"message":"Cannot read property 'build' of undefined","name":"TypeError","stack":"TypeError: Cannot read property 'build' of undefined\n    at Container.<anonymous> (/opt/probo/lib/Container.js:193:39)\n    at /opt/probo/lib/ContainerManager.js:359:47\n    at next (native)\n    at onFulfilled (/opt/probo/node_modules/co/index.js:65:19)\n    at bound (domain.js:280:14)\n    at runBound (domain.js:293:12)\n    at runCallback (timers.js:574:20)\nFrom previous event:\n    at next (/opt/probo/node_modules/co/index.js:100:51)\n    at onFulfilled (/opt/probo/node_modules/co/index.js:69:7)\n    at /opt/probo/node_modules/co/index.js:54:5\n    at co (/opt/probo/node_modules/co/index.js:50:10)\n    at /opt/probo/lib/safeco.js:9:12\n    at Immediate.<anonymous> (/opt/probo/lib/ContainerManager.js:354:5)\n    at runCallback (timers.js:576:20)\n    at tryOnImmediate (timers.js:554:5)\n    at processImmediate [as _immediateCallback] (timers.js:533:5)"},"msg":"Uncaught co Exception","time":"2016-10-11T19:31:16.615Z","src":{"file":"/opt/probo/lib/ContainerManager.js","line":142},"v":0}
```

Let me know if anyone has thoughts or ideas on this one. Happy to test plausiable solutions!

-Brian

**Comment by sk33lz** [comment]
This should be fixed. Please re-open if this is not the case.

**Comment by sk33lz** [comment]
Reopening due to related issues. Might need another ticket once we figure it all out.

**Comment by brianyoungblood** [comment]
The json output from the error I'm getting for the 404 message is:

`{"name":"proxy","hostname":"li171-190","pid":32027,"req_id":"4b555e2d-1634-4917-9b88-47a34399a0e2","level":30,"dest":{"dest":"577235f846800000","build":"577235f846800000"},"msg":"using dest:","time":"2017-08-09T21:16:12.100Z","src":{"file":"/opt/probo-proxy/lib/proxy.js","line":99},"v":0}
{"name":"proxy","hostname":"li171-190","pid":32027,"component":"config","level":20,"msg":"returning pre-loaded config","time":"2017-08-09T21:16:12.100Z","src":{"file":"/opt/probo-proxy/lib/config.js","line":60,"func":"load"},"v":0}
{"name":"proxy","hostname":"li171-190","pid":32027,"req_id":"4b555e2d-1634-4917-9b88-47a34399a0e2","component":"proxy-lookup","level":20,"msg":"proxy lookup cache MISS: 577235f846800000","time":"2017-08-09T21:16:12.101Z","src":{"file":"/opt/probo-proxy/lib/proxy-lookup.js","line":81,"func":"lookup"},"v":0}
{"name":"probo","hostname":"li171-190","pid":31957,"component":"container-manager","id":"fc174ab8-9dcf-4efb-bada-2127e4ebe1d0","req_id":"f8633250-7d47-11e7-a365-4f6c1710f203","level":30,"req":{"method":"POST","url":"/container/proxy?build=577235f846800000&dest=577235f846800000","headers":{"host":"localhost:3020","content-length":"0","connection":"close"},"remoteAddress":"127.0.0.1","remotePort":46042},"msg":"REQUEST","time":"2017-08-09T21:16:12.149Z","src":{"file":"/opt/probo/lib/ContainerManager.js","line":114},"v":0}
{"name":"proxy","hostname":"li171-190","pid":32027,"req_id":"4b555e2d-1634-4917-9b88-47a34399a0e2","level":50,"err":{"message":"Build id is invalid","name":"Error","stack":"Error: Build id is invalid\n    at handleNonSuccessResponse (/opt/probo-proxy/lib/proxy-lookup.js:153:18)\n    at Request._callback (/opt/probo-proxy/lib/proxy-lookup.js:114:7)\n    at Request.self.callback (/opt/probo-proxy/node_modules/request/request.js:187:22)\n    at emitTwo (events.js:106:13)\n    at Request.emit (events.js:191:7)\n    at Request.<anonymous> (/opt/probo-proxy/node_modules/request/request.js:1048:10)\n    at emitOne (events.js:96:13)\n    at Request.emit (events.js:188:7)\n    at IncomingMessage.<anonymous> (/opt/probo-proxy/node_modules/request/request.js:969:12)\n    at emitNone (events.js:91:20)"},"msg":"Error looking up proxy information","time":"2017-08-09T21:16:12.164Z","src":{"file":"/opt/probo-proxy/lib/proxy.js","line":42,"func":"respondWithLookupError"},"v":0}
{"name":"probo","hostname":"li171-190","pid":31957,"component":"http","audit":true,"level":30,"remoteAddress":"127.0.0.1","remotePort":46042,"req_id":"f8633250-7d47-11e7-a365-4f6c1710f203","req":{"method":"POST","url":"/container/proxy?build=577235f846800000&dest=577235f846800000","headers":{"host":"localhost:3020","content-length":"0","connection":"close"},"httpVersion":"1.1","trailers":{},"version":"*","timers":{"parseAccept":235,"readBody":27,"parseBody":7,"parseQueryString":139,"handler-4":183,"bunyan":80,"handler-6":12965,"bound ":2957}},"res":{"statusCode":400,"headers":{"x-request-id":"fc174ab8-9dcf-4efb-bada-2127e4ebe1d0","content-type":"application/json","content-length":52},"trailer":false},"latency":17,"_audit":true,"msg":"handled: 400","time":"2017-08-09T21:16:12.165Z","src":{"file":"/opt/probo/node_modules/restify/lib/plugins/audit.js","line":94,"func":"audit"},"v":0}
{"name":"probo","hostname":"li171-190","pid":31957,"component":"container-manager","id":"fc174ab8-9dcf-4efb-bada-2127e4ebe1d0","req_id":"f8633250-7d47-11e7-a365-4f6c1710f203","level":30,"ip":"127.0.0.1","method":"POST","url":"/container/proxy?build=577235f846800000&dest=577235f846800000","req":{"method":"POST","url":"/container/proxy?build=577235f846800000&dest=577235f846800000","headers":{"host":"localhost:3020","content-length":"0","connection":"close"},"remoteAddress":"127.0.0.1","remotePort":46042},"res":{"statusCode":400,"header":"HTTP/1.1 400 Bad Request\r\nx-request-id: fc174ab8-9dcf-4efb-bada-2127e4ebe1d0\r\nContent-Type: application/json\r\nContent-Length: 52\r\nDate: Wed, 09 Aug 2017 21:16:12 GMT\r\nConnection: close\r\n\r\n"},"duration":17.503163,"msg":"127.0.0.1 - 2017-08-09T21:16:12.166Z POST /container/proxy?build=577235f846800000&dest=577235f846800000 HTTP/1.1 400","time":"2017-08-09T21:16:12.166Z","src":{"file":"/opt/probo/node_modules/probo-request-logger/index.js","line":40,"func":"responseSent"},"v":0}
{"name":"proxy","hostname":"li171-190","pid":32027,"req_id":"6a6ec5b0-fe4f-4bdf-bd55-aa104751fe47","level":30,"dest":{"dest":"577235f846800000","build":"577235f846800000"},"msg":"using dest:","time":"2017-08-09T21:16:14.344Z","src":{"file":"/opt/probo-proxy/lib/proxy.js","line":99},"v":0}
{"name":"proxy","hostname":"li171-190","pid":32027,"component":"config","level":20,"msg":"returning pre-loaded config","time":"2017-08-09T21:16:14.344Z","src":{"file":"/opt/probo-proxy/lib/config.js","line":60,"func":"load"},"v":0}
{"name":"proxy","hostname":"li171-190","pid":32027,"req_id":"6a6ec5b0-fe4f-4bdf-bd55-aa104751fe47","component":"proxy-lookup","level":20,"msg":"proxy lookup cache MISS: 577235f846800000","time":"2017-08-09T21:16:14.345Z","src":{"file":"/opt/probo-proxy/lib/proxy-lookup.js","line":81,"func":"lookup"},"v":0}
{"name":"probo","hostname":"li171-190","pid":31957,"component":"container-manager","id":"c495c576-52d0-4e76-9fc3-03bbd8c08e36","req_id":"f9b331f0-7d47-11e7-a365-4f6c1710f203","level":30,"req":{"method":"POST","url":"/container/proxy?build=577235f846800000&dest=577235f846800000","headers":{"host":"localhost:3020","content-length":"0","connection":"close"},"remoteAddress":"127.0.0.1","remotePort":46056},"msg":"REQUEST","time":"2017-08-09T21:16:14.351Z","src":{"file":"/opt/probo/lib/ContainerManager.js","line":114},"v":0}
{"name":"probo","hostname":"li171-190","pid":31957,"component":"http","audit":true,"level":30,"remoteAddress":"127.0.0.1","remotePort":46056,"req_id":"f9b331f0-7d47-11e7-a365-4f6c1710f203","req":{"method":"POST","url":"/container/proxy?build=577235f846800000&dest=577235f846800000","headers":{"host":"localhost:3020","content-length":"0","connection":"close"},"httpVersion":"1.1","trailers":{},"version":"*","timers":{"parseAccept":253,"readBody":23,"parseBody":5,"parseQueryString":158,"handler-4":303,"bunyan":155,"handler-6":391,"bound ":307}},"res":{"statusCode":400,"headers":{"x-request-id":"c495c576-52d0-4e76-9fc3-03bbd8c08e36","content-type":"application/json","content-length":52},"trailer":false},"latency":1,"_audit":true,"msg":"handled: 400","time":"2017-08-09T21:16:14.351Z","src":{"file":"/opt/probo/node_modules/restify/lib/plugins/audit.js","line":94,"func":"audit"},"v":0}
{"name":"probo","hostname":"li171-190","pid":31957,"component":"container-manager","id":"c495c576-52d0-4e76-9fc3-03bbd8c08e36","req_id":"f9b331f0-7d47-11e7-a365-4f6c1710f203","level":30,"ip":"127.0.0.1","method":"POST","url":"/container/proxy?build=577235f846800000&dest=577235f846800000","req":{"method":"POST","url":"/container/proxy?build=577235f846800000&dest=577235f846800000","headers":{"host":"localhost:3020","content-length":"0","connection":"close"},"remoteAddress":"127.0.0.1","remotePort":46056},"res":{"statusCode":400,"header":"HTTP/1.1 400 Bad Request\r\nx-request-id: c495c576-52d0-4e76-9fc3-03bbd8c08e36\r\nContent-Type: application/json\r\nContent-Length: 52\r\nDate: Wed, 09 Aug 2017 21:16:14 GMT\r\nConnection: close\r\n\r\n"},"duration":1.342477,"msg":"127.0.0.1 - 2017-08-09T21:16:14.352Z POST /container/proxy?build=577235f846800000&dest=577235f846800000 HTTP/1.1 400","time":"2017-08-09T21:16:14.352Z","src":{"file":"/opt/probo/node_modules/probo-request-logger/index.js","line":40,"func":"responseSent"},"v":0}
{"name":"proxy","hostname":"li171-190","pid":32027,"req_id":"6a6ec5b0-fe4f-4bdf-bd55-aa104751fe47","level":50,"err":{"message":"Build id is invalid","name":"Error","stack":"Error: Build id is invalid\n    at handleNonSuccessResponse (/opt/probo-proxy/lib/proxy-lookup.js:153:18)\n    at Request._callback (/opt/probo-proxy/lib/proxy-lookup.js:114:7)\n    at Request.self.callback (/opt/probo-proxy/node_modules/request/request.js:187:22)\n    at emitTwo (events.js:106:13)\n    at Request.emit (events.js:191:7)\n    at Request.<anonymous> (/opt/probo-proxy/node_modules/request/request.js:1048:10)\n    at emitOne (events.js:96:13)\n    at Request.emit (events.js:188:7)\n    at IncomingMessage.<anonymous> (/opt/probo-proxy/node_modules/request/request.js:969:12)\n    at emitNone (events.js:91:20)"},"msg":"Error looking up proxy information","time":"2017-08-09T21:16:14.354Z","src":{"file":"/opt/probo-proxy/lib/proxy.js","line":42,"func":"respondWithLookupError"},"v":0}`

**Comment by ElusiveMind** [comment]
I believe I have diagnosed this problem. If a buildID (data.id) is not provided to Server.prototype.storeInDB (line 1059 of ContainerManager.js) then it creates one. However, it does not create one that passes the regex at line ~637 of the same file. Therefore when the proxy passes the buildID in the incorrect format, it fails and produces the invalid buildID error. 

I have exchanged the existing code which relies on FlakeID with some code that generates the groups of hex needed to pass the regex. This was done to produce as little collateral damage to the project as possible so this can be submitted as a pull request to the project.

My code looks like this:

`data.id = `
`      crypto.randomBytes(4).toString('hex') + '-' +`
`      crypto.randomBytes(2).toString('hex') + '-' +`
`      crypto.randomBytes(2).toString('hex') + '-' +`
`      crypto.randomBytes(2).toString('hex') + '-' +`
`      crypto.randomBytes(6).toString('hex');`

I spent a few hours researching a "better" way to do this but without doing a bunch of string manipulation and producing hashes which was unreliable (I read) to which I would split and rejoin (slow) I though this might be the best solution. I submit it for your approval.
